odoo/addons/base/models/ir_cron.py

```diff
@@ -520,6 +563,29 @@ def _notifydb(self):
             cr.execute(query, [self.env.cr.dbname])
         _logger.debug("cron workers notified")
 
+    def _create_progress(self, cron_id, *, timed_out=False):
```

Comment:
Suggestion: `cron_id` should be `self.id`, and it could return `self` with the progress context:
```py
cron_with_context = cron._create_progress()
...
cron_with_context._notify_progress(done, remaining)
```
This way, the first method provides what makes the second method work, and this context hack is no longer mixed with the rest of the code.

Reply:
Maybe you need an extra method to discard the timeout on the current progress:
```py
cron_with_context._complete_progress()
```

Reply:
But then you have to read from context to get the progress infos which is a "context hack" has well ?

(Or search in a loop but....)